### ISSUE TITLE
Add tests for std::pipe

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3830]
`std` now has a [`pipe` module](https://doc.rust-lang.org/nightly/std/pipe/index.html). We should add a test in Miri to see which OS APIs are required for this.
- On Unix, AFAIK this uses `pipe2`. Miri has basic support for that but does not support blocking, so properly supporting `std::pipe` is blocked on the socketpair part of https://github.com/rust-lang/miri/issues/3665.
- On Windows, for now it's okay to ignore the test like we do for many other OS APIs. But of course if someone wants to implement this, they can give it a go. :) This will require someone reasonably familiar with Windows APIs since the Miri team does not have the expertise for that.

**Comment by tiif** [comment]
It works :) [Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4ad39c33c25128ab9d688038774d9553)

Do we still want to add a test for this?

**Comment by RalfJung** [comment]
Yes, we want to be sure we support whatever flags std is setting, at least on Linux and if possible macOS.

**Comment by tiif** [comment]
It couldn't be run on macOS and Windows. 

macOS: ``ioctl``
```rust
error: unsupported operation: can't call foreign function `ioctl` on OS `macos`
   --> /home/tiif/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/sys/pal/unix/fd.rs:510:17
    |
510 |             cvt(libc::ioctl(self.as_raw_fd(), libc::FIOCLEX))?;
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `ioctl` on OS `macos`
```

Windows: ``CreatePipe```
```rust
error: unsupported operation: can't call foreign function `CreatePipe` on OS `windows`
  --> /home/tiif/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/sys/anonymous_pipe/windows.rs:17:24
   |
17 |     let ret = unsafe { c::CreatePipe(&mut read_pipe, &mut write_pipe, ptr::null_mut(), 0) };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `CreatePipe` on OS `windows`
```

**Comment by RalfJung** [comment]
Yeah the Windows failure is expected, but it'd be good to get macOS working.

**Comment by tiif** [comment]
The ``libc::FIOCLEX`` seems to be the ``close-on-exec`` flag

> [FIOCLEX](https://man.openbsd.org/ioctl.2#FIOCLEX)
    Set close-on-exec flag. The file will be closed when [execve(2)](https://man.openbsd.org/execve.2) is invoked.


which should be easy to support since we don't have ``exec``, so nothing needs to be done there?

**Comment by tiif** [comment]
Created an issue for that https://github.com/rust-lang/miri/issues/4126#issue-2772629702.

**Comment by tiif** [comment]
This is completed via #4133.
